application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/GetTrialLicenseListener.java

```diff
@@ -65,10 +66,14 @@ public String getName()
     @Override
     public void onEvent(Event event, Object source, Object data)
     {
-        DefaultInstalledExtension extension = (DefaultInstalledExtension) source;
+        List<ExtensionId> extensions = ((JobFinishedEvent) event).getRequest().getProperty("extensions");
+        // Retrieve license updates to be sure that we don't override an existing license.
+        trialLicenseGenerator.updateLicenses();
 
-        if (trialLicenseGenerator.canGenerateTrialLicense(extension.getId())) {
-            trialLicenseGenerator.generateTrialLicense(extension.getId());
+        for (ExtensionId extensionId : extensions) {
+            if (trialLicenseGenerator.canGenerateTrialLicense(extensionId)) {
```

Comment:
Although it's not common, we can have a free extension that depends on a paid extension. AFAIK, the list of extensions you get from the job request represents only the top level extensions that the user has requested to be installed explicitly. It doesn't include the transitive dependencies. So you need to go through the dependency tree (starting from these top level extensions) to look for licensed extensions. Of course, you don't go "down" when you find a licensed extension (its dependencies will be covered).